### ISSUE TITLE
Feature request: Fix dimension convention for masks in transformer

### ISSUE THREAD

**Issue by malmaud** [pytorch/pytorch#25132]
In PyTorch 1.2,

For `nn.Transformer`, the `src` dimensions correspond to `S, N, E` (sequence position, batch index, embedding feature), but the `src_key_padding_mask` dimensions correspond to `N, S`. That is, the batch is the second dimension of `src` but the first dimension of `src_key_padding_mask` (and other masks). I find this pretty confusing and inconsistent: either all parameters should be batch-first or they all should be batch-second.

Relatedly, there should probably be an option analogous to the LSTM's `batch_first` argument.

**Comment by soumith** [comment]
cc: @zhangguanheng66 @myleott

**Comment by zhangguanheng66** [comment]
Thanks @malmaud

Within the MultiheadAttention module, `key_padding_mask` is applied before the `softmax` layer. At this point, `attn_output_weights` has the `bsz` dim before the `src_len` dim. IMO, it makes sense to have `key_padding_mask` in the shape of (`bsz src_len`), without the extra `transpose` ops.
